grpcreflect.go

```diff
@@ -408,74 +392,31 @@ func (n *staticNames) Names() []string {
 	return n.names
 }
 
-// hackedResolver provides hacks to workaround the fact that connect-grpcreflect-go
-// and connect-grpchealth-go use hacked "connectext.grpc..." packages in the linked-in
-// descriptors (to avoid init-time panics due to conflicts, in case calling code also
-// links in the gRPC runtime's versions of these services).
-type hackedResolver struct {
-	resolver atomic.Pointer[protodesc.Resolver]
-	init     sync.Once
-}
-
-func resolverHackForConnectext() *hackedResolver {
-	res := &hackedResolver{}
-	var delegate protodesc.Resolver = protoregistry.GlobalFiles
-	res.resolver.Store(&delegate)
-	return res
-}
-
-func (r *hackedResolver) FindFileByPath(path string) (protoreflect.FileDescriptor, error) {
-	res := *r.resolver.Load()
-	file, err := res.FindFileByPath(path)
-	if err != nil {
-		if res == protoregistry.GlobalFiles && contains(mangledFileNames, path) {
-			// we need to create "unmangled" versions of the services
-			// that connect libs have to mangle
-			res = r.doInit(res)
-			file, err = res.FindFileByPath(path)
-			if err == nil {
-				return file, nil
-			}
-		}
-		return nil, err
+// resolverHackForConnectext returns a resolver that can successfully resolve the descriptors
+// for the gRPC health and reflection services. We need a work-around since this repo (and the
+// connect-grpchealth-go repo) use "hacked" services that have a "connectext." package prefix.
+// We don't use the "authoritative" packages for these descriptors because they depend on the
+// gRPC runtime (ew!). We add a special prefix to the packages to avoid an init-time panic from
+// duplicate registrations, in the event that the calling application _also_ imports the gRPC
+// versions.
+//
+// This works by serving embedded descriptors (from "services.bin") for items not found in
+// protoregistry.GlobalFiles. The only thing in the embedded descriptors are for the health
+// and reflection services.
+func resolverHackForConnectext(data []byte) protodesc.Resolver {
+	var resolver protodesc.Resolver = protoregistry.GlobalFiles
+	var fileSet descriptorpb.FileDescriptorSet
+	if err := proto.Unmarshal(data, &fileSet); err != nil {
+		return resolver
 	}
-	return file, nil
-}
-
-func (r *hackedResolver) FindDescriptorByName(name protoreflect.FullName) (protoreflect.Descriptor, error) {
-	res := *r.resolver.Load()
-	desc, err := res.FindDescriptorByName(name)
+	files, err := protodesc.NewFiles(&fileSet)
 	if err != nil {
-		if res == protoregistry.GlobalFiles && contains(mangledServiceNames, string(name)) {
-			// we need to create "unmangled" versions of the services
-			// that connect libs have to mangle
-			res = r.doInit(res)
-			desc, err = res.FindDescriptorByName(name)
-			if err == nil {
-				return desc, nil
-			}
-		}
-		return nil, err
+		return resolver
```

Comment:
I'm uncomfortable ignoring errors here - don't they indicate some critical incompatibility between the embedded binary data, the schemas, and `protodesc`? If the owner of `main` is relying on reflection to find the health and reflection services, we don't want to fail at runtime with no explanation.

Could we instead return `(protodesc.Resolver, error)`, capture the error as a global, and return it from `combinedResolver` if we ever need to look up symbols in `.second`?

Reply:
> don't they indicate some critical incompatibility between the embedded binary data, the schemas, and protodesc?

Yes, but it's also a condition that would certainly be caught during CI testing.

> Could we instead return `(protodesc.Resolver, error)`, capture the error as a global, and return it from `combinedResolver` if we ever need to look up symbols in `.second`?

Sure.

Reply:
Done. 0df528356d1aca3c1b57b7cb9da1b4d2f907016e